### PR TITLE
Notification settings UI refresh

### PR DESCRIPTION
**Pull request by turt2live**
Fixes https://github.com/vector-im/element-web/issues/17782
**Requires https://github.com/matrix-org/matrix-js-sdk/pull/1778**

----

***Code reviewer***: This is a fairly large change. I highly recommend reviewing commit-by-commit, and then changing your diff view for the rewrite of the `<Notifications>` component.

----

image
image

### CODE REVIEW

File: src/components/views/settings/Notifications.tsx [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+                });
+
+                // XXX: Do we need this block from the previous component?
+                /*
+                    // if there was a rule which we couldn't parse, add it to the external list
+                    if (rule && !vectorState) {
+                        rule.description = ruleDefinition.description;
+                        self.state.externalPushRules.push(rule);
+                    }
+                 */
```
Comment (turt2live) [inline src/components/views/settings/Notifications.tsx:225]: ftr we don't really do anything with this in the old component, so I'm tempted to just forget about it.

File: src/components/views/settings/Notifications.tsx [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+                const rule = defaultRules.others[i];
+                const ruleDescription = otherRulesDescriptions[rule.rule_id];
+
+                // Show enabled default rules that was modified by the user
+                if (ruleDescription && rule.enabled && !rule.default) {
+                    rule.description = ruleDescription;
+                    self.state.externalPushRules.push(rule);
+                }
+            }
+         */
```
Comment (turt2live) [inline src/components/views/settings/Notifications.tsx:267]: I have no idea what this does in the old component.

**Comment by SimonBrandner** [comment]
Wow, this looks great :smile: Though I think it might be nicer if the email notifs had their own section but that's just a nitpick

File: src/components/views/settings/Notifications.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/views/settings/Notifications.tsx:225]: So they would do something but it has to be driven by the server, where the server injects its own `.something.custom` push rules into the scope as a user can't create `.` push rules

**Review by gaelledel — APPROVED** [review]
Amazing!

File: src/components/views/settings/Notifications.tsx (hunk 1 above)
Comment (turt2live) [reply src/components/views/settings/Notifications.tsx:225]: sure, but all we did with them was list them. Not sure we really need to do that in our UI anymore...

File: src/components/views/elements/TagComposer.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        this.props.onAdd(this.state.newTag);
+        this.setState({ newTag: "" });
+    };
+
+    private onRemove = (tag: string) => {
+        // We probably don't need to proxy this, but for
+        // sanity of `this` we'll do so anyways.
+        this.props.onRemove(tag);
+    };
```
Comment (t3chguy) [inline src/components/views/elements/TagComposer.tsx:66]: given that you bind this on use it doesn't need to be an arrow func

File: src/components/views/settings/Notifications.tsx [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+        return preparedNewState;
+    }
+
+    private async refreshPushers(): Promise<Partial<IState>> {
+        return { ...(await MatrixClientPeg.get().getPushers()) };
+    }
+
+    private async refreshThreepids(): Promise<Partial<IState>> {
+        return { ...(await MatrixClientPeg.get().getThreePids()) };
+    }
```
Comment (t3chguy) [inline src/components/views/settings/Notifications.tsx:249]: why are the objects being exploded here given that you're about to pluck their values into a new object via Object.assign anyhow?

**Review by t3chguy — APPROVED** [review]
Looks sane to me otherwise!!

File: res/css/views/settings/_Notifications.scss [hunk 5]
```diff
@@ -14,82 +14,79 @@ See the License for the specific language governing permissions and
...
+
+    .mx_UserNotifSettings_pushRulesTable {
+        width: calc(100% + 12px); // +12px to line up center of 'Noisy' column with toggle switches
+        table-layout: fixed;
+        border-collapse: collapse;
+        border-spacing: 0;
+        margin-top: 40px;
+
+        tr > th {
+            font-weight: 600; // semi bold
```
[the remark this answers is not in the document]
Comment (turt2live) [reply res/css/views/settings/_Notifications.scss:28]: TIL these are variables

File: src/components/views/settings/Notifications.tsx (hunk 4 above)
Comment (turt2live) [reply src/components/views/settings/Notifications.tsx:249]: Because I like complex code? Honestly not sure wtf I did 🙃

**Pull request merged**